File: mobile/scripts/qa-mobile-fidelity.mjs
/**
 * qa-mobile-fidelity.mjs — Playwright QA for the 2026-06-11 mobile-fidelity pass.
 *
 * Verifies mobile runs the IDENTICAL desktop experiences (not flat fallbacks)
 * across iPhone 12 / Galaxy S22 / iPhone SE, + desktop no-regression + a
 * reduced-motion pass. Server must already be on :5050.
 *
 * Resolves Playwright from the npx cache (not a repo dep) so package.json stays
 * untouched. Run:  node mobile/scripts/qa-mobile-fidelity.mjs
 */
import { createRequire } from "node:module";
const require = createRequire(import.meta.url);

// Resolve the npx-cached playwright module.
const PW_CACHE = "C:/Users/art1/AppData/Local/npm-cache/_npx/5e2e484947874241/node_modules/playwright";
let chromium;
try {
  ({ chromium } = require(PW_CACHE));
} catch {
  ({ chromium } = require("playwright"));
}

const BASE = "http://localhost:5050";
const MOBILE = [["iPhone12", 390, 844], ["GalaxyS22", 360, 780], ["iPhoneSE", 375, 667]];
const results = [];
const log = (tag, ok, detail = "") => {
  results.push([ok ? "PASS" : "FAIL", tag, detail]);
  console.log(`  [${ok ? "PASS" : "FAIL"}] ${tag}${detail ? " — " + detail : ""}`);
};
const settle = async (page) => { await page.waitForLoadState("networkidle").catch(() => {}); await page.waitForTimeout(400); };

async function testHome(page, label, w, reduced = false) {
  console.log(`\n=== HOME @ ${label} (${w}px)${reduced ? " [reduced-motion]" : ""} ===`);
  await page.goto(BASE + "/", { waitUntil: "domcontentloaded" });
  await settle(page);

  const layers = await page.evaluate(() => {
    const q = (s) => document.querySelector(s);
    const ev = q(".hero-static__layer--events > img");
    const rs = q(".hero-static__layer--resort > img");
    const op = (el) => (el ? parseFloat(getComputedStyle(el).opacity) : -1);
    const l = q(".hero-static__cta-label");
    let pill = "none";
    if (l) { const c = getComputedStyle(l); pill = (c.position === "absolute" && parseInt(c.width) <= 2) ? "sr-only" : "VISIBLE-PILL"; }
    return {
      base: !!q(".hero-static__layer--base"), gamos: !!q(".hero-static__layer--gamos"),
      desert: !!q(".hero-static__layer--desert"),
      eventsImg: op(ev), eventsW: ev ? ev.getBoundingClientRect().width : 0, resortImg: op(rs),
      tapEvents: !!q(".hero-static__tap--events"), pill,
    };
  });
  log(`hero/5-layers/${label}`, layers.base && layers.gamos && layers.desert, "base+gamos+desert present");
  log(`hero/events-bitmap-visible/${label}`, layers.eventsImg === 1 && layers.eventsW > 10, `opacity=${layers.eventsImg} w=${Math.round(layers.eventsW)}`);
  log(`hero/resort-bitmap-visible/${label}`, layers.resortImg === 1, `opacity=${layers.resortImg}`);
  log(`hero/no-pill/${label}`, layers.pill === "sr-only", `cta-label=${layers.pill}`);
  if (w <= 768) {
    const tap = await page.evaluate(() => {
      const t = document.querySelector(".hero-static__tap--events");
      if (!t) return { ok: false };
      const r = t.getBoundingClientRect(); const c = getComputedStyle(t);
      return { ok: true, w: r.width, h: r.height, href: t.getAttribute("href"), z: c.zIndex };
    });
    log(`hero/tap-overlay->=48px/${label}`, tap.ok && tap.w >= 48 && tap.h >= 48, `w=${Math.round(tap.w || 0)} h=${Math.round(tap.h || 0)} href=${tap.href} z=${tap.z}`);
  }

  const ov = await page.evaluate(() => ({ sw: document.documentElement.scrollWidth, cw: document.documentElement.clientWidth }));
  log(`no-h-overflow/${label}`, ov.sw <= ov.cw + 2, `scrollWidth=${ov.sw} clientWidth=${ov.cw}`);

  await page.evaluate(() => document.querySelector("#lounge")?.scrollIntoView());
  await page.waitForTimeout(700);
  const lounge = await page.evaluate(() => {
    const ring = document.querySelector("[data-lounge-ring]"); const stage = document.querySelector("[data-lounge-stage]");
    if (!ring || !stage) return { ok: false };
    return { ok: true, ringT: getComputedStyle(ring).transform, persp: getComputedStyle(stage).perspective, stageT: getComputedStyle(stage).transform };
  });
  if (reduced) {
    log(`lounge/RM-grid/${label}`, lounge.ok && (lounge.ringT === "none" || lounge.ringT === "matrix(1, 0, 0, 1, 0, 0)"), `ringT=${lounge.ringT}`);
  } else {
    const hasRotate = lounge.ok && lounge.ringT && lounge.ringT.includes("matrix") && lounge.ringT !== "none";
    const perspOn = lounge.ok && lounge.persp !== "none";
    log(`lounge/3D-ring-live/${label}`, hasRotate && perspOn, `ringT=${(lounge.ringT || "?").slice(0, 28)} persp=${lounge.persp}`);
    if (w <= 768) log(`lounge/stage-scaled/${label}`, (lounge.stageT || "").includes("matrix"), `stageT=${(lounge.stageT || "none").slice(0, 28)}`);
  }

  if (w <= 768) {
    const man = await page.evaluate(() => { const c = document.querySelector(".culinary canvas[data-manifest-url]"); return c ? c.dataset.manifestUrl : "no-canvas"; });
    log(`culinary/mobile-manifest/${label}`, (man || "").includes("culinary-mobile"), `manifest=${man}`);
  }

  const shab = await page.evaluate(() => {
    const lay = document.querySelector(".shabbat__layout"); const stg = document.querySelector(".shabbat__stage");
    return { layD: lay ? getComputedStyle(lay).display : "none", stgD: stg ? getComputedStyle(stg).display : "none" };
  });
  log(`shabbat/not-flattened/${label}`, shab.layD !== "contents" && shab.stgD !== "contents", `layout=${shab.layD} stage=${shab.stgD}`);
}

async function testPress(page, label) {
  console.log(`\n=== PRESS @ ${label} ===`);
  await page.goto(BASE + "/press/", { waitUntil: "domcontentloaded" });
  await settle(page);
  const pr = await page.evaluate(() => {
    const host = document.querySelector("[data-press-shader]");
    return { host: !!host, canvases: host ? host.querySelectorAll("canvas").length : 0, rows: document.querySelectorAll(".press-card, [data-stagger] > *").length };
  });
  log(`press/shader-mounts/${label}`, pr.host && pr.canvases >= 1, `host=${pr.host} canvases=${pr.canvases}`);
  log(`press/rows/${label}`, pr.rows > 0, `rows=${pr.rows}`);
}

async function testHalls(page, label) {
  console.log(`\n=== HALLS oasis @ ${label} (touch) ===`);
  await page.goto(BASE + "/halls/dist/oasis/", { waitUntil: "domcontentloaded" });
  await page.waitForTimeout(1800);
  const h = await page.evaluate(() => { const c = document.querySelector("canvas"); return { canvas: !!c, w: c ? c.width : 0, h: c ? c.height : 0 }; });
  log(`halls/webgl-canvas/${label}`, h.canvas && h.w > 0, `canvas ${h.w}x${h.h}`);
}

async function testRooms(page, label) {
  console.log(`\n=== ROOMS wall @ ${label} (touch) ===`);
  await page.goto(BASE + "/rooms/dist/", { waitUntil: "domcontentloaded" });
  await page.waitForTimeout(1800);
  const r = await page.evaluate(() => { const c = document.querySelector("canvas"); return { canvas: !!c, w: c ? c.width : 0 }; });
  log(`rooms/wall-canvas/${label}`, r.canvas && r.w > 0, `canvas w=${r.w}`);
}

const browser = await chromium.launch({ headless: true });
for (const [label, w, hgt] of MOBILE) {
  const ctx = await browser.newContext({ viewport: { width: w, height: hgt }, isMobile: true, hasTouch: true, deviceScaleFactor: 2 });
  const page = await ctx.newPage();
  await testHome(page, label, w);
  if (label === "iPhone12") { await testPress(page, label); await testHalls(page, label); await testRooms(page, label); }
  await ctx.close();
}
{ // desktop no-regression
  const ctx = await browser.newContext({ viewport: { width: 1440, height: 900 } });
  const page = await ctx.newPage(); await testHome(page, "Desktop1440", 1440); await ctx.close();
}
{ // reduced-motion mobile
  const ctx = await browser.newContext({ viewport: { width: 390, height: 844 }, isMobile: true, hasTouch: true, reducedMotion: "reduce" });
  const page = await ctx.newPage(); await testHome(page, "iPhone12-RM", 390, true); await ctx.close();
}
await browser.close();

console.log("\n" + "=".repeat(60));
const passed = results.filter((r) => r[0] === "PASS").length;
const failed = results.filter((r) => r[0] === "FAIL");
console.log(`TOTAL: ${passed}/${results.length} passed`);
if (failed.length) { console.log("\nFAILURES:"); for (const [, tag, d] of failed) console.log(`  FAIL ${tag} — ${d}`); }
else console.log("ALL CHECKS PASSED");

File: mobile/scripts/qa-mobile-fidelity.py
#!/usr/bin/env python3
"""
qa-mobile-fidelity.py — Playwright QA for the 2026-06-11 mobile-fidelity pass.

Verifies mobile now runs the IDENTICAL desktop experiences (not flat fallbacks)
across iPhone 12 / Galaxy S22 / iPhone SE, plus a desktop no-regression pass and
a reduced-motion pass. Server must already be running on :5050.

Run (server managed separately, already up):
    python mobile/scripts/qa-mobile-fidelity.py
"""
from playwright.sync_api import sync_playwright

BASE = "http://localhost:5050"
MOBILE = [("iPhone12", 390, 844), ("GalaxyS22", 360, 780), ("iPhoneSE", 375, 667)]
results = []


def log(tag, ok, detail=""):
    mark = "PASS" if ok else "FAIL"
    results.append((mark, tag, detail))
    print(f"  [{mark}] {tag}{(' — ' + detail) if detail else ''}")


def settle(page):
    page.wait_for_load_state("networkidle")
    page.wait_for_timeout(400)


def test_home(page, label, w, reduced=False):
    print(f"\n=== HOME @ {label} ({w}px){' [reduced-motion]' if reduced else ''} ===")
    page.goto(BASE + "/", wait_until="domcontentloaded")
    settle(page)

    # --- HERO: real 5-layer composition, bitmaps visible (not pills), tap overlays ---
    layers = page.evaluate("""() => {
      const q = s => document.querySelector(s);
      const ev = q('.hero-static__layer--events > img');
      const rs = q('.hero-static__layer--resort > img');
      const cs = el => el ? getComputedStyle(el) : null;
      const evc = cs(ev), rsc = cs(rs);
      return {
        base: !!q('.hero-static__layer--base'),
        gamos: !!q('.hero-static__layer--gamos'),
        eventsImg: ev ? parseFloat(evc.opacity) : -1,
        eventsW: ev ? ev.getBoundingClientRect().width : 0,
        resortImg: rs ? parseFloat(rsc.opacity) : -1,
        desert: !!q('.hero-static__layer--desert'),
        tapEvents: !!q('.hero-static__tap--events'),
        tapResort: !!q('.hero-static__tap--resort'),
        pill: (() => { const l = q('.hero-static__cta-label');
          if (!l) return 'none'; const c = getComputedStyle(l);
          return (c.position === 'absolute' && parseInt(c.width) <= 2) ? 'sr-only' : 'VISIBLE-PILL'; })(),
      };
    }""")
    log(f"hero/5-layers/{label}", layers["base"] and layers["gamos"] and layers["desert"],
        f"base+gamos+desert present")
    log(f"hero/events-bitmap-visible/{label}", layers["eventsImg"] == 1 and layers["eventsW"] > 10,
        f"events img opacity={layers['eventsImg']} w={round(layers['eventsW'])}")
    log(f"hero/resort-bitmap-visible/{label}", layers["resortImg"] == 1,
        f"resort img opacity={layers['resortImg']}")
    log(f"hero/no-pill/{label}", layers["pill"] == "sr-only", f"cta-label={layers['pill']}")
    if w <= 768:
        tap = page.evaluate("""() => {
          const t = document.querySelector('.hero-static__tap--events');
          if (!t) return {ok:false};
          const r = t.getBoundingClientRect(); const c = getComputedStyle(t);
          return {ok:true, w:r.width, h:r.height, href:t.getAttribute('href'),
                  z:c.zIndex, bg:c.backgroundColor};
        }""")
        log(f"hero/tap-overlay-≥48px/{label}", tap.get("ok") and tap["w"] >= 48 and tap["h"] >= 48,
            f"w={round(tap.get('w',0))} h={round(tap.get('h',0))} href={tap.get('href')} z={tap.get('z')}")

    # --- no horizontal overflow anywhere on the page (lounge ring fit etc.) ---
    ov = page.evaluate("() => ({sw: document.documentElement.scrollWidth, cw: document.documentElement.clientWidth})")
    log(f"no-h-overflow/{label}", ov["sw"] <= ov["cw"] + 2, f"scrollWidth={ov['sw']} clientWidth={ov['cw']}")

    # --- LOUNGE: real 3D ring (rotateY transform on ring, perspective on stage) ---
    page.evaluate("() => document.querySelector('#lounge')?.scrollIntoView()")
    page.wait_for_timeout(700)
    lounge = page.evaluate("""() => {
      const ring = document.querySelector('[data-lounge-ring]');
      const stage = document.querySelector('[data-lounge-stage]');
      if (!ring || !stage) return {ok:false};
      return {ok:true, ringT: getComputedStyle(ring).transform,
              persp: getComputedStyle(stage).perspective,
              stageT: getComputedStyle(stage).transform};
    }""")
    if reduced:
        # RM → core grid: ring transform should be none (JS short-circuits)
        log(f"lounge/RM-grid/{label}", lounge.get("ok") and lounge["ringT"] in ("none", "matrix(1, 0, 0, 1, 0, 0)"),
            f"ringT={lounge.get('ringT')}")
    else:
        has_rotate = lounge.get("ok") and ("matrix3d" in lounge["ringT"] or "matrix" in lounge["ringT"]) and lounge["ringT"] != "none"
        persp_on = lounge.get("ok") and lounge["persp"] != "none"
        log(f"lounge/3D-ring-live/{label}", has_rotate and persp_on,
            f"ringT={lounge.get('ringT','?')[:28]} persp={lounge.get('persp')}")
        if w <= 768:
            log(f"lounge/stage-scaled/{label}", "matrix" in (lounge.get("stageT") or ""),
                f"stageT={lounge.get('stageT','none')[:28]}")

    # --- CULINARY: mobile manifest active on phones ---
    if w <= 768:
        man = page.evaluate("""() => {
          const c = document.querySelector('.culinary canvas[data-manifest-url]');
          return c ? c.dataset.manifestUrl : 'no-canvas';
        }""")
        log(f"culinary/mobile-manifest/{label}", "culinary-mobile" in (man or ""), f"manifest={man}")

    # --- SHABBAT: layout NOT display:contents (pin survives) ---
    shab = page.evaluate("""() => {
      const lay = document.querySelector('.shabbat__layout');
      const stg = document.querySelector('.shabbat__stage');
      return {layD: lay ? getComputedStyle(lay).display : 'none',
              stgD: stg ? getComputedStyle(stg).display : 'none'};
    }""")
    log(f"shabbat/not-flattened/{label}", shab["layD"] != "contents" and shab["stgD"] != "contents",
        f"layout.display={shab['layD']} stage.display={shab['stgD']}")


def test_press(page, label):
    print(f"\n=== PRESS @ {label} ===")
    page.goto(BASE + "/press/", wait_until="domcontentloaded")
    settle(page)
    pr = page.evaluate("""() => {
      const host = document.querySelector('[data-press-shader]');
      const canvases = host ? host.querySelectorAll('canvas').length : 0;
      const rows = document.querySelectorAll('.press-card, [data-stagger] > *').length;
      return {host: !!host, canvases, rows};
    }""")
    log(f"press/shader-mounts/{label}", pr["host"] and pr["canvases"] >= 1,
        f"host={pr['host']} canvases={pr['canvases']}")
    log(f"press/rows/{label}", pr["rows"] > 0, f"rows={pr['rows']}")


def test_halls(page, label):
    print(f"\n=== HALLS events @ {label} (touch) ===")
    page.goto(BASE + "/halls/dist/events/", wait_until="domcontentloaded")
    page.wait_for_timeout(1500)
    h = page.evaluate("""() => {
      const c = document.querySelector('canvas');
      return {canvas: !!c, w: c ? c.width : 0, h: c ? c.height : 0,
              err: window.__lastError || null};
    }""")
    log(f"halls/webgl-canvas/{label}", h["canvas"] and h["w"] > 0, f"canvas {h['w']}x{h['h']}")


def test_rooms(page, label):
    print(f"\n=== ROOMS wall @ {label} (touch) ===")
    page.goto(BASE + "/rooms/dist/", wait_until="domcontentloaded")
    page.wait_for_timeout(1500)
    r = page.evaluate("""() => {
      const c = document.querySelector('canvas');
      return {canvas: !!c, w: c ? c.width : 0};
    }""")
    log(f"rooms/wall-canvas/{label}", r["canvas"] and r["w"] > 0, f"canvas w={r['w']}")


with sync_playwright() as p:
    browser = p.chromium.launch(headless=True)

    # Mobile matrix (touch + coarse pointer)
    for (label, w, hgt) in MOBILE:
        ctx = browser.new_context(viewport={"width": w, "height": hgt},
                                  is_mobile=True, has_touch=True, device_scale_factor=2)
        page = ctx.new_page()
        test_home(page, label, w)
        if label == "iPhone12":
            test_press(page, label)
            test_halls(page, label)
            test_rooms(page, label)
        ctx.close()

    # Desktop no-regression
    ctx = browser.new_context(viewport={"width": 1440, "height": 900})
    page = ctx.new_page()
    test_home(page, "Desktop1440", 1440)
    ctx.close()

    # Reduced-motion (mobile)
    ctx = browser.new_context(viewport={"width": 390, "height": 844},
                              is_mobile=True, has_touch=True, reduced_motion="reduce")
    page = ctx.new_page()
    test_home(page, "iPhone12-RM", 390, reduced=True)
    ctx.close()

    browser.close()

print("\n" + "=" * 60)
passed = sum(1 for r in results if r[0] == "PASS")
failed = [r for r in results if r[0] == "FAIL"]
print(f"TOTAL: {passed}/{len(results)} passed")
if failed:
    print("\nFAILURES:")
    for _, tag, detail in failed:
        print(f"  FAIL {tag} — {detail}")
else:
    print("ALL CHECKS PASSED")
